### ISSUE TITLE
Multi-file target matching without output statement.

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SOS#450]
Separated from #444 

```
parameter: fileset1 = ['{}.txt'.format(i) for i in range(2)]
parameter: fileset2 = ['{}.txt'.format(i) for i in range(2,4)]
parameter: fileset3 = ['{}.txt'.format(i) for i in range(4,6)]
parameter: fileset4 = ['{}.txt'.format(i) for i in range(6,10)]
parameter: fileset5 = ['{}.txt'.format(i) for i in range(10,14)]

[main]
parameter: output_files = []
input: for_each=['output_files']
output: output_files[_index]
task: concurrent = True
run:
  touch ${_output}

[joe_a: provides=fileset1]
output: fileset1
sos_run('main', output_files = fileset1)

[john_a: provides=fileset2]
depends: fileset1
output: fileset2
sos_run('main', output_files = fileset2)

[john_b: provides=fileset3]
depends: fileset1
output: fileset3
sos_run('main', output_files = fileset3)

[mark_a: provides=fileset4]
depends: fileset1, fileset2
output: fileset4
sos_run('main', output_files = fileset4)

[mark_b: provides=fileset5]
depends: fileset1, fileset3
output: fileset5
sos_run('main', output_files = fileset5)

[default]
depends: fileset4, fileset5
```

would fail without additional `output` statement. SoS should handle it or give warning messages.

**Comment by gaow** [comment]
I have a related question: currently I'm in a situation where each `fileset?` is a very very long list of files. It can take huge amount of time for SoS to build the execution path. Would it be helpful / harmful if I write this instead:

```
[joe_a: provides=fileset1[0] ]
output: fileset1
sos_run('main', output_files = fileset1)
```
where I explicitly ask for one single file from the list to represent the output, knowing that all steps do not "provide" overlaps?

**Comment by BoPeng** [comment]
Yes you can, but things should be optimizable if the dag does not grow. I will check if I get some time. Is very long about 1k or 10k? It is possible that your filesystem is acting up with many small files.

**Comment by BoPeng** [comment]
Part of the problem is that sos needs to calculate md5 for all tracked files, that can take a while.

**Comment by gaow** [comment]
> Part of the problem is that sos needs to calculate md5 for all tracked files

I've changed my code to only `provide` one representative. It does speed it up a lot. The total number of files involved (that requires md5 check otherwise) is 46,800. The DAG structure is not complicated. It is similar to this one below but just having many parallel branches happening at the same time.

![res](https://cloud.githubusercontent.com/assets/917985/23174093/e5a5339a-f820-11e6-8f78-8c6c1f2b85fa.png)

**Comment by BoPeng** [comment]
Can we design some sort of `bag` target or `dir` target to track a bunch of files or directory without md5 of the exact content? For example, 

```
files(["{x}.txt" for x in range(10000)"])
```

would be considered as a single target. The exists of `files` could be existence of all files, but `md5` would be simply md5 of filenames (or cheap information about file size). Similarly, 

```
files('a_dir')
```
would just check all the files.

**Comment by gaow** [comment]
Great -- this could be very helpful for this context: but how about other context where this can help? I'll open a separate ticket for this.
